src/vm/database/marf.rs

```diff
@@ -0,0 +1,125 @@
+use std::path::PathBuf;
+
+use vm::errors::{ InterpreterError, InterpreterResult as Result, IncomparableError };
+use vm::database::{KeyValueStorage, SqliteConnection};
+use chainstate::stacks::index::marf::MARF;
+use chainstate::stacks::index::{MARFValue, Error as MarfError};
+use chainstate::stacks::index::storage::{TrieFileStorage};
+use chainstate::burn::BlockHeaderHash;
+
+/// The MarfedKV struct is used to wrap a MARF data structure and side-storage
+///   for use as a K/V store for ClarityDB or the AnalysisDB.
+/// The Clarity VM and type checker do not "know" to begin/commit the block they are currently processing:
+///   each instantiation of the VM simply executes one transaction. So the block handling
+///   loop will need to invoke these two methods (begin + commit) outside of the context of the VM.
+///   NOTE: Clarity will panic if you try to execute it from a non-initialized MarfedKV context.
+///   (See: vm::tests::with_marfed_environment()) 
+pub struct MarfedKV {
+    chain_tip: BlockHeaderHash,
```

Comment:
Why are we tracking the `chain_tip` here, in addition to inside the `MARF` instance?  Does the `MARF` need to be modified to make it easier to track this information in one place?

Reply:
I'm not sure -- the reason this is done is to allow `get`s to query the current chain tip in situations where `begin()` shouldn't have been called (when we're not actively evaluating a block) -- at the moment, that's only in the `check` routine in `clarity.rs`.

Reply:
Ah.  I think there's something missing from the user-facing interface -- when `get`-ing anything, the user will either need to specify a chain tip hash, or the system should look up the tip of the longest fork and use that as the default.

Reply:
Regardless, this struct should not be storing the current chain tip.  The fork table ultimately tells you what the chain tips are, and maybe there needs to be a fork table method for getting the longest chain's tip.

Reply:
No, I think here we'd want to always use the "current chain tip" as set in the MarfedKV struct (or via calling `begin()` and always using `.open_chain_tip()`) -- a Clarity context evaluates a single transaction, meaning that the caller will set the chain tip (which may not necessarily be the longest chain tip, rather its whichever chain tip the current block is building off of), and then it should always be the same throughout the execution of the Clarity context.

Reply:
So what the would mean is that `get()` would default to the open chain tip (not the longest chain tip)

Reply:
Gotcha.  I want to make sure that if `.begin()` gets called, then `self.chain_tip` is always equal to `.open_chain_tip()`.  Is there a way the code can enforce this?

Reply:
Sure -- I changed `MarfedKV.begin()` to do:

```
...
        self.chain_tip = self.marf.get_open_chain_tip()
            .unwrap().clone();
...
```

Which enforces that property -- the MARF otherwise cannot be begun, since its a private variable, and this code will panic anyways if the MARF was previously opened.